### ISSUE TITLE
Unable to reproduce results before and after a study is deepcopied

### ISSUE THREAD

**Issue by Y-oHr-N** [optuna/optuna#914]
The behavior is different between 0.19.0 and 1.x.x.

## Environment

- Python version: 3.6.5
- OS: Windows 10

## Examples

```python
import copy

import optuna


def objective(trial):
    x = trial.suggest_uniform('x', -10, 10)

    return (x - 2) ** 2


study = optuna.create_study(sampler=optuna.samplers.TPESampler(seed=0))
study.optimize(objective, n_trials=1)

study = optuna.create_study(sampler=optuna.samplers.TPESampler(seed=0))
study = copy.deepcopy(study)
study.optimize(objective, n_trials=1)
```

### v0.19.0

```
[I 2020-02-12 02:37:41,933] Finished trial#0 resulted in value: 1.0480229520791995. Current best value is 1.0480229520791995 with parameters: {'x': 0.9762700785464951}.
[I 2020-02-12 02:37:41,990] Finished trial#0 resulted in value: 1.0480229520791995. Current best value is 1.0480229520791995 with parameters: {'x': 0.9762700785464951}.
```

### v1.0.0

```
[I 2020-02-12 02:38:44,010] Finished trial#0 resulted in value: 1.0480229520791995. Current best value is 1.0480229520791995 with parameters: {'x': 0.9762700785464951}.
[I 2020-02-12 02:38:44,067] Finished trial#0 resulted in value: 46.90638356377688. Current best value is 46.90638356377688 with parameters: {'x': 8.848823516763801}.
```

### v1.1.0

```
[I 2020-02-12 02:39:39,168] Finished trial#0 resulted in value: 1.0480229520791995. Current best value is 1.0480229520791995 with parameters: {'x': 0.9762700785464951}.
[I 2020-02-12 02:39:39,226] Finished trial#0 resulted in value: 2.756825098011473. Current best value is 2.756825098011473 with parameters: {'x': 3.6603689644207016}.
```

**Comment by c-bata** [comment]
It seems a cause of this problem is the following commit.
https://github.com/optuna/optuna/commit/3f01ca234d91dd5281e784f9c0f5f124f897f275

**Comment by Y-oHr-N** [comment]
I have read https://github.com/optuna/optuna/pull/692#issuecomment-558510300.

> But I don't think the approach to initialize the RandomSampler is the main topic of this PR, we can keep the current implementation and discuss it after introducing the joblib.

Are there any plans to fix it in the future to ensure the reproducibility?

**Comment by sile** [comment]
Hi @Y-oHr-N, thank you for reporting this issue.
I created a draft PR #967 to address this problem (this PR introduces additional overhead to samplers so it might not be accepted though).

**Comment by Y-oHr-N** [comment]
Thank you for your quick response. I hope the PR is successfully merged.

**Comment by sile** [comment]
@Y-oHr-N Since this problem seems to have been solved by #968, I'll close this issue. Feel free to reopen it if you still have any problems.
